### PR TITLE
off-chain workers htg - transactions

### PR DESCRIPTION
**Pull request by jimmychu0807**
@samelamin This is the draft PR on off-chain workers. I will work on the transactions one.
Feel free to pick any of the rest to work on. Hopefully we can complete this PR before Christmas.

@sacha-l @NukeManDan fyi, we are working on this PR.

Part 1: #684 - transactions
Part 2: #724 - local storage
Part 3: #729 - off-chain indexing

### CODE REVIEW

**Comment by samelamin** [comment]
Thanks for this @jimmychu0807 

ill work on HTTP requests

**Comment by jimmychu0807** [comment]
@sacha-l @NukeManDan @lisa-parity 

The how-to guide for using off-chain worker for sending three type of transactions is completed.

Lisa: Please review my english and the style as appropriate. You could just directly overwrite on top of my branch.

Sacha & Dan: Please review the content and any formatting. 

Thank you!

@samelamin 

Sam: I have commented out the http requests and the rest of the chapters on the sidebar that are not ready to be shown. Feel free to comment them back in when you make a PR on *Making HTTP Requests* section.

I will work on the *Local Storage* section next. Thank.

---
Last but not least, happy holiday to you all 🎉 🎄 🎅

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+        log::error!("No local account available");
+      }
+    }
+  }
+  ```
+
+  In the above code snippet, we fetch the `signer` and then call `send_unsigned_transaction()`.
+  It accepts two function closures. The first one returns the payload to be used, and the second
+  one returns the on-chain call with payload and signature passed in. This call returns a
+  `Option<(Account<T>, Result<(), ()>)>` result type, which carry the meaning:
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:436]:
```suggestion
  `Option<(Account<T>, Result<(), ()>)>` result type to allow for the following results:
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+
+    * `None`: no account is available for sending transaction
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:438]:
```suggestion
    * `None` if no account is available for sending the transaction.
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+    * `Some((account, Ok(())))`: transaction is successfully sent
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:439]:
```suggestion
    * `Some((account, Ok(())))` if the transaction is successfully sent.
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+    * `Some((account, Err(())))`: error occured when sending the transaction
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:440]:
```suggestion
    * `Some((account, Err(())))` if an error occurs when sending the transaction.
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+
+3. Lastly, you may have a more complex `ValidateUnsigned` logic to check the provided signature
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:442]:
```suggestion
You can implement more complex logic for `ValidateUnsigned` trait, if needed.
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+  does match with the public key signing the payload. You can do that with the following logics
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:443]:
```suggestion
  For example, to check whether a provided signature matches the public key used to sign the payload, you can implement the `ValidateUnsigned` trait using code similar to the following:
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+  in the `ValidateUnsigned` implementation.
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:444]:
```suggestion
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+          }
+          valid_tx(b"unsigned_extrinsic_with_signed_payload".to_vec())
+        },
+        _ => InvalidTransaction::Call.into(),
+      }
+    }
+  }
+  ```
+
+  We have seen most of the above code before. The new code is that we use
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:475]:
```suggestion
  This example uses [`SignedPayload`](https://paritytech.github.io/substrate/latest/frame_system/offchain/trait.SignedPayload.html?search=SignedPayload) to verify that the public key in the payload has the same signature as the one provided.
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+  [`SignedPayload`](https://paritytech.github.io/substrate/latest/frame_system/offchain/trait.SignedPayload.html?search=SignedPayload)
+  to verify the payload (which contains the public key) has the same signature as the one provided.
+
+A full example of [the off-chain function call](https://github.com/paritytech/substrate/blob/monthly-2021-12/frame/examples/offchain-worker/src/lib.rs#L505-L532)
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:479]:
```suggestion
You can see a full example in the [off-chain function call](https://github.com/paritytech/substrate/blob/monthly-2021-12/frame/examples/offchain-worker/src/lib.rs#L505-L532) and [`ValidateUnsigned` implementation](https://github.com/paritytech/substrate/blob/monthly-2021-12/frame/examples/offchain-worker/src/lib.rs#L308-L318).
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+and [`ValidateUnsigned` implementation](https://github.com/paritytech/substrate/blob/monthly-2021-12/frame/examples/offchain-worker/src/lib.rs#L308-L318)
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:480]:
```suggestion
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+can be seen.
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:481]:
```suggestion
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+
+---
+
+Whew! This is a long chapter. But congratulation! By now you should be able to use off-chain workers
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:485]:
```suggestion
You have now seen how you can use off-chain workers to send data for on-chain storage using:
```

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 13]
```diff
@@ -0,0 +1,499 @@
...
+to send data on-chain using:
```
Comment (lisa-parity) [inline v3/how-to-guides/04-ocw/a-transactions/index.mdx:486]:
```suggestion
```

**Review by lisa-parity — COMMENTED** [review]
This was a little hard for me to follow in some spots. I tried to reduce some of the redundancy and steps more active.

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 14]
```diff
@@ -0,0 +1,499 @@
...
+---
+
+<Objectives data={[
+  {
+    title: 'Goal',
+    description: `Learn how to perform the following actions from off-chain workers:
+- Signed transactions
+- Unsigned transactions
+- Unsigned transactions with signed payload`
+  },
```
[the remark this answers is not in the document]
Comment (jimmychu0807) [reply v3/how-to-guides/04-ocw/a-transactions/index.mdx:20]: I think we need the bracket, don't we?

[the remark this answers is not in the document]
Comment (jimmychu0807) [reply v3/how-to-guides/04-ocw/a-transactions/index.mdx:20]: I see. You are suggesting to remove the `Use Cases` paragraph.

File: v3/how-to-guides/04-ocw/a-transactions/index.mdx [hunk 15]
```diff
@@ -0,0 +1,499 @@
...
+
+  ```rust
+  fn offchain_worker(block_number: T::BlockNumber) {
+    let signer = Signer::<T, T::AuthorityId>::all_accounts();
+
+    // ...
+  }
+  ```
+
+  Although it is calling `all_accounts()`, we will only inject one account into this pallet later
```
[the remark this answers is not in the document]
Comment (jimmychu0807) [reply v3/how-to-guides/04-ocw/a-transactions/index.mdx]: I have re-written this sentence.

**Comment by jimmychu0807** [comment]
@lisa-parity the article reads a lot better after your suggestion. Thank you for this!

I have re-written the sentence that read ambiguous on `all_accounts()`, see if it is more clear now. Thanks.

**Comment by jimmychu0807** [comment]
I learned:

1. One sentence per line
2. Remove redundant "we" by using command tone (`Do this ...`, `Call this...`, `Write the following code...`).

**Review by sacha-l — APPROVED** [review]
Added a couple edits, looks good to me now.

**Pull request merged**